### ISSUE TITLE
[question] How the 'pi/logstd' file is created?

### ISSUE THREAD

**Issue by FabioPINO** [hill-a/stable-baselines#952]
I am writing to have a simple clarification about the standard deviation in the continuos action probability. How is the file logstd.npy created? Why is it saved in a file if the action probability depends on the policy values?  I tried to construct a custom policy with a deterministic action selection (code below) following the tutorial reported on your website, I'd like to know how the action is selected in this case?
```
        mean = linear(pi_latent_vector, 'pi', self.size, init_scale=init_scale, init_bias=init_bias)
        logstd = tf.get_variable(name='pi/logstd', shape=[1, self.size], initializer=tf.zeros_initializer())
        pdparam = tf.concat([mean, mean * 0.0 + logstd], axis=1)
q_values = linear(vf_latent_vector, 'q', self.size, init_scale=init_scale, init_bias=init_bias)

```
Custom policy
```
    class CustomPolicy(ActorCriticPolicy):
        def __init__(self, sess, ob_space, ac_space, n_env, n_steps, n_batch, reuse=False, **kwargs):
            super(CustomPolicy, self).__init__(sess, ob_space, ac_space, n_env, n_steps, n_batch, reuse=reuse, scale=True)

            with tf.variable_scope("model", reuse=reuse):
                activ = tf.nn.tanh

                extracted_features = self.processed_obs

                pi_h = extracted_features
                for i, layer_size in enumerate([128,128,128]):
                    pi_h = activ(tf.layers.dense(pi_h, layer_size, name='pi_fc' + str(i)))
                pi_latent = pi_h

                vf_h = extracted_features
                for i, layer_size in enumerate([32, 32]):
                    vf_h = activ(tf.layers.dense(vf_h, layer_size, name='vf_fc' + str(i)))
                
                value_fn = (tf.layers.dense(vf_h, 1, name='vf'))
                vf_latent = vf_h

                self._proba_distribution, self._policy, self.q_value = \
                    self.pdtype.proba_distribution_from_latent(pi_latent, vf_latent, init_scale=0.01)

            self._value_fn = value_fn
            self._setup_init()

        def step(self, obs, state=None, mask=None, deterministic=True):
            if deterministic:
                action, value, neglogp = self.sess.run([self.deterministic_action, self.value_flat, self.neglogp],
                                                       {self.obs_ph: obs})
            else:
                action, value, neglogp = self.sess.run([self.action, self.value_flat, self.neglogp],
                                                       {self.obs_ph: obs})
            return action, value, self.initial_state, neglogp

        def proba_step(self, obs, state=None, mask=None):
            return self.sess.run(self.policy_proba, {self.obs_ph: obs})

        def value(self, obs, state=None, mask=None):
            return self.sess.run(self.value_flat, {self.obs_ph: obs})
```

**Comment by Miffyli** [comment]
No files are being created, the `tf.get_variable` simply queries for a variable name in the tensorflow graph and returns it. As I have understood it, standard deviation is fixed and not predicted by the agent (@araffin could you confirm?). Deterministic action takes the mean directly without sampling from a squashed Gaussian.

**Comment by araffin** [comment]
>As I have understood it, standard deviation is fixed and not predicted by the agent (@araffin could you confirm?). Deterministic action takes the mean directly without sampling from a squashed Gaussian.

yes and yes ;)

The standard deviation of the Gaussian is created in the distribution.
Best is to take a look at SB3: https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/distributions.py#L140

It could in theory depends on the state (like for SAC), see a longer answer I already gave here: https://github.com/hill-a/stable-baselines/issues/652

**Comment by FabioPINO** [comment]
Thank you for your prompt replies and your clarifications @araffin @Miffyli!
Is it possible to change the std value manually? I am a bit confused, if the std is constant throughout the learning, how the exploration/exploitation is handled? I though that the std was progressively reduced to concentrate the action selection on the mean value provided by the actor. So, how is the exploration actually managed and what is the role of the ent_coeff?

**Comment by araffin** [comment]
>Is it possible to change the std value manually? I am a bit confused, if the std is constant throughout the learning, how the exploration/exploitation is handled? 

You need to do a bit of math to understand how it works (this [paper](http://people.idsia.ch/~juergen/paladyn2010.pdf) is a good introduction). You can derive the objective function with respect to the std (assuming that your policy follows a Gaussian distribution), and hence can update the std using gradient descent (that's what is done currently).
Alternatively, you can also manually set and update that value (that was done for the Humanoid environment in the original paper).

>I though that the std was progressively reduced to concentrate the action selection on the mean value provided by the actor. 

If the optimal policy is deterministic, that's what will happen.

>what is the role of the ent_coeff?

The entropy coefficient will favor exploration by preventing the std to decrease too soon (again you need a bit a math to see that formally), but it is usually not needed for continuous control environments (when the std of the Gaussian is defined).

**Comment by FabioPINO** [comment]
Thank you @araffin  for your thorough reply! Is there a particular reason behind the choice of using a gradient descent method for the std update in stable baselines? By reading the papers and other resources I noticed the std can be computed based on the states or on the policy, as the mean.

**Comment by araffin** [comment]
>By reading the papers and other resources I noticed the std can be computed based on the states or on the policy, as the mean.

Could you elaborate?
